esx_service/vmdk_ops.py

```diff
@@ -794,78 +851,57 @@ def disk_attach(vmdk_path, vm):
+            msg = "Failed to place new disk - out of disk slots due to no slot to add a new controller"
             logging.error(msg + " VM=%s", vm.config.uuid)
             return err(msg)
 
-        # find empty bus slot for the controller:
-        taken = set([c.busNumber for c in controllers])
-        avail = set(range(0, max_scsi_controllers)) - taken
-
-        key = avail.pop()  # bus slot
-        controller_key = key + offset_from_bus_number
-        disk_slot = 0
-        controller_spec = vim.VirtualDeviceConfigSpec(
-            operation='add',
-            device=vim.ParaVirtualSCSIController(key=controller_key,
-                                                 busNumber=key,
-                                                 sharedBus='noSharing', ), )
-        # changes spec content goes here
-        pvscsi_change = []
-        pvscsi_change.append(controller_spec)
-        spec = vim.vm.ConfigSpec()
-        spec.deviceChange = pvscsi_change
+        logging.info("Trying to add one PVSCSI controller")
+        
+        controller_key, ret_err = add_pvscsi_controller(vm, controllers, max_scsi_controllers, 
+                                                        offset_from_bus_number)
 
-        try:
-            wait_for_tasks(si, [vm.ReconfigVM_Task(spec=spec)])
-        except vim.fault.VimFault as ex:
-            msg=("Failed to add PVSCSI Controller: %s", ex.msg)
-            return err(msg)
+        if (ret_err is not None):
+            return ret_err    
+            
         # Find the controller just added
         devices = vm.config.hardware.device
         pvsci = [d for d in devices
                  if type(d) == vim.ParaVirtualSCSIController and
                  d.key == controller_key]
-        pci_slot_number = get_controller_pci_slot(vm, pvsci,
+        pci_slot_number = get_controller_pci_slot(vm, pvsci[0],
                                                   offset_from_bus_number)
+        logging.debug("Added a PVSCSI controller, controller_key=%d pci_slot_number=%d",
+                      controller_key, pci_slot_number)
 
     # Find a slot on the controller, issue attach task and wait for completion
-    if not disk_slot:
-        taken = set([dev.unitNumber
-                     for dev in devices
-                     if type(dev) == vim.VirtualDisk and dev.controllerKey ==
-                     controller_key])
-        # search in 15 slots, with unit_number 7 reserved for scsi controller
-        availSlots = (set(range(0, 7)) | set(range(8, 16))) - taken
-
-        if len(availSlots) == 0:
-            msg = "Failed to place new disk - out of disk slots"
-            logging.error(msg + " VM=%s", vm.config.uuid)
-            return err(msg)
-
-        disk_slot = availSlots.pop()
-        logging.debug("controller_key = %d slot = %d", controller_key, disk_slot)
-
+    if disk_slot is None:
+       msg = "Failed to place new disk - out of disk slots"
```

Comment:
Use same error message as above?
"Failed to attach new volume - The maximum number of supported volumes of 60 has been reached."